File: ui/lib/models/ipc_protocol.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';

part 'ipc_protocol.freezed.dart';
part 'ipc_protocol.g.dart';

/// 纯IPC协议 - 与Python端完全一致
/// 完全摒弃HTTP概念，使用IPC特定的成功/失败标志

/// IPC错误信息
@freezed
class IPCError with _$IPCError {
  const factory IPCError({
    required String code,
    required String message,
    Map<String, dynamic>? details,
  }) = _IPCError;

  factory IPCError.fromJson(Map<String, dynamic> json) =>
      _$IPCErrorFromJson(json);
}

/// IPC元数据
@freezed
class IPCMetadata with _$IPCMetadata {
  const factory IPCMetadata({
    required String timestamp,
    @JsonKey(name: 'request_id') String? requestId,
  }) = _IPCMetadata;

  factory IPCMetadata.fromJson(Map<String, dynamic> json) =>
      _$IPCMetadataFromJson(json);
}

/// 纯IPC响应格式
@freezed
class IPCResponse with _$IPCResponse {
  const factory IPCResponse({
    required bool success,
    dynamic data,  // 允许任何类型的data，包括Map、List等
    IPCError? error,
    IPCMetadata? metadata,
  }) = _IPCResponse;

  factory IPCResponse.fromJson(Map<String, dynamic> json) =>
      _$IPCResponseFromJson(json);

  /// 创建成功响应
  factory IPCResponse.success({
    dynamic data,
    String? requestId,
  }) {
    return IPCResponse(
      success: true,
      data: data,
      metadata: IPCMetadata(
        timestamp: DateTime.now().toIso8601String(),
        requestId: requestId,
      ),
    );
  }

  /// 创建错误响应
  factory IPCResponse.failure({
    required String errorCode,
    required String message,
    Map<String, dynamic>? details,
    String? requestId,
  }) {
    return IPCResponse(
      success: false,
      error: IPCError(
        code: errorCode,
        message: message,
        details: details,
      ),
      metadata: IPCMetadata(
        timestamp: DateTime.now().toIso8601String(),
        requestId: requestId,
      ),
    );
  }
}

/// IPC请求格式
@freezed
class IPCRequest with _$IPCRequest {
  const factory IPCRequest({
    required String method,
    required String path,
    Map<String, dynamic>? data,
    @JsonKey(name: 'query_params') Map<String, dynamic>? queryParams,
    @JsonKey(name: 'path_params') Map<String, String>? pathParams,
    @JsonKey(name: 'request_id') String? requestId,
  }) = _IPCRequest;

  factory IPCRequest.fromJson(Map<String, dynamic> json) =>
      _$IPCRequestFromJson(json);

  /// 创建GET请求
  factory IPCRequest.get({
    required String path,
    Map<String, dynamic>? queryParams,
    String? requestId,
  }) {
    return IPCRequest(
      method: 'GET',
      path: path,
      queryParams: queryParams,
      requestId: requestId,
    );
  }

  /// 创建POST请求
  factory IPCRequest.post({
    required String path,
    Map<String, dynamic>? data,
    String? requestId,
  }) {
    return IPCRequest(
      method: 'POST',
      path: path,
      data: data,
      requestId: requestId,
    );
  }

  /// 创建PUT请求
  factory IPCRequest.put({
    required String path,
    Map<String, dynamic>? data,
    String? requestId,
  }) {
    return IPCRequest(
      method: 'PUT',
      path: path,
      data: data,
      requestId: requestId,
    );
  }
}

/// 连接器运行状态枚举 - 与Python/C++完全一致
enum ConnectorRunningState {
  @JsonValue('stopped')
  stopped,

  @JsonValue('starting')
  starting,

  @JsonValue('running')
  running,

  @JsonValue('stopping')
  stopping,

  @JsonValue('error')
  error;

  /// 获取显示文本
  String get displayName {
    switch (this) {
      case ConnectorRunningState.stopped:
        return '已停止';
      case ConnectorRunningState.starting:
        return '启动中';
      case ConnectorRunningState.running:
        return '运行中';
      case ConnectorRunningState.stopping:
        return '停止中';
      case ConnectorRunningState.error:
        return '错误';
    }
  }

  /// 获取颜色
  int get color {
    switch (this) {
      case ConnectorRunningState.stopped:
        return 0xFF666666; // 灰色
      case ConnectorRunningState.starting:
        return 0xFF2196F3; // 蓝色
      case ConnectorRunningState.running:
        return 0xFF4CAF50; // 绿色
      case ConnectorRunningState.stopping:
        return 0xFFFF9800; // 橙色
      case ConnectorRunningState.error:
        return 0xFFF44336; // 红色
    }
  }

  /// 从字符串创建
  static ConnectorRunningState fromString(String state) {
    switch (state.toLowerCase()) {
      case 'stopped':
        return ConnectorRunningState.stopped;
      case 'starting':
        return ConnectorRunningState.starting;
      case 'running':
        return ConnectorRunningState.running;
      case 'stopping':
        return ConnectorRunningState.stopping;
      case 'error':
        return ConnectorRunningState.error;
      default:
        return ConnectorRunningState.stopped;
    }
  }
}

/// 连接器状态信息 V2 - 使用新的状态模型
@freezed
class ConnectorStatusV2 with _$ConnectorStatusV2 {
  const factory ConnectorStatusV2({
    @JsonKey(name: 'connector_id') required String connectorId,
    @JsonKey(name: 'display_name') required String displayName,
    @Default(true) bool enabled, // 是否启用（基本状态）
    @JsonKey(name: 'running_state')
    @Default(ConnectorRunningState.stopped)
        ConnectorRunningState runningState, // 运行状态
    @JsonKey(name: 'is_installed') @Default(true) bool isInstalled, // 虚拟字段
    @JsonKey(name: 'is_healthy') @Default(false) bool isHealthy, // 是否健康
    @JsonKey(name: 'should_be_running') @Default(false) bool shouldBeRunning, // 是否应该运行
    @JsonKey(name: 'process_id') int? processId,
    @JsonKey(name: 'last_heartbeat') String? lastHeartbeat,
    @JsonKey(name: 'data_count') @Default(0) int dataCount,
    @JsonKey(name: 'last_activity') String? lastActivity,
    @JsonKey(name: 'error_message') String? errorMessage,
    @JsonKey(name: 'error_code') String? errorCode,
  }) = _ConnectorStatusV2;

  factory ConnectorStatusV2.fromJson(Map<String, dynamic> json) =>
      _$ConnectorStatusV2FromJson(json);

  /// 创建新连接器的默认状态
  factory ConnectorStatusV2.newConnector({
    required String connectorId,
    required String displayName,
  }) {
    return ConnectorStatusV2(
      connectorId: connectorId,
      displayName: displayName,
      enabled: true,
      runningState: ConnectorRunningState.stopped,
      isInstalled: true,
    );
  }

  /// 创建禁用连接器状态
  factory ConnectorStatusV2.disabled({
    required String connectorId,
    required String displayName,
  }) {
    return ConnectorStatusV2(
      connectorId: connectorId,
      displayName: displayName,
      enabled: false,
      runningState: ConnectorRunningState.stopped,
      isInstalled: true,
    );
  }
}

/// 常用IPC错误码
class IPCErrorCode {
  // 连接和认证错误
  static const String connectionFailed = 'IPC_CONNECTION_FAILED';
  static const String authRequired = 'IPC_AUTH_REQUIRED';
  static const String authFailed = 'IPC_AUTH_FAILED';
  
  // 请求处理错误
  static const String invalidRequest = 'IPC_INVALID_REQUEST';
  static const String missingParameter = 'IPC_MISSING_PARAMETER';
  static const String requestTimeout = 'IPC_REQUEST_TIMEOUT';
  
  // 连接器相关错误
  static const String connectorNotFound = 'CONNECTOR_NOT_FOUND';
  static const String connectorConfigInvalid = 'CONNECTOR_CONFIG_INVALID';
  static const String connectorStartFailed = 'CONNECTOR_START_FAILED';
  static const String connectorStopFailed = 'CONNECTOR_STOP_FAILED';
  
  // 系统错误
  static const String internalError = 'INTERNAL_ERROR';
  static const String serviceUnavailable = 'SERVICE_UNAVAILABLE';
  static const String resourceNotFound = 'RESOURCE_NOT_FOUND';
}

File: daemon/services/ipc_protocol.py
"""
纯IPC通信协议 - 完全独立于HTTP概念
定义统一的数据交换格式和错误处理体系
"""

import json
import uuid
from dataclasses import asdict, dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional, Union


class IPCErrorCode(Enum):
    """IPC特定错误码体系"""

    # 连接和认证错误
    CONNECTION_FAILED = "IPC_CONNECTION_FAILED"
    AUTH_REQUIRED = "IPC_AUTH_REQUIRED"
    AUTH_FAILED = "IPC_AUTH_FAILED"
    CLIENT_DISCONNECTED = "IPC_CLIENT_DISCONNECTED"

    # 请求处理错误
    INVALID_REQUEST = "IPC_INVALID_REQUEST"
    MISSING_PARAMETER = "IPC_MISSING_PARAMETER"
    INVALID_PARAMETER = "IPC_INVALID_PARAMETER"
    REQUEST_TIMEOUT = "IPC_REQUEST_TIMEOUT"

    # 连接器相关错误
    CONNECTOR_NOT_FOUND = "CONNECTOR_NOT_FOUND"
    CONNECTOR_CONFIG_INVALID = "CONNECTOR_CONFIG_INVALID"
    CONNECTOR_START_FAILED = "CONNECTOR_START_FAILED"
    CONNECTOR_STOP_FAILED = "CONNECTOR_STOP_FAILED"
    CONNECTOR_ALREADY_RUNNING = "CONNECTOR_ALREADY_RUNNING"
    CONNECTOR_NOT_RUNNING = "CONNECTOR_NOT_RUNNING"

    # 系统错误
    INTERNAL_ERROR = "INTERNAL_ERROR"
    SERVICE_UNAVAILABLE = "SERVICE_UNAVAILABLE"
    DATABASE_ERROR = "DATABASE_ERROR"

    # 资源错误
    RESOURCE_NOT_FOUND = "RESOURCE_NOT_FOUND"
    RESOURCE_LOCKED = "RESOURCE_LOCKED"
    INSUFFICIENT_PERMISSIONS = "INSUFFICIENT_PERMISSIONS"


@dataclass
class IPCMessage:
    """IPC基础消息类"""

    method: str
    path: str
    data: Dict[str, Any] = None
    headers: Dict[str, str] = None
    query_params: Dict[str, Any] = None

    def __post_init__(self):
        if self.data is None:
            self.data = {}
        if self.headers is None:
            self.headers = {}
        if self.query_params is None:
            self.query_params = {}

    def to_json(self) -> str:
        """序列化为JSON字符串"""
        return json.dumps(asdict(self), ensure_ascii=False)

    @classmethod
    def from_json(cls, json_str: str) -> "IPCMessage":
        """从JSON字符串反序列化"""
        data = json.loads(json_str)
        return cls(**data)


@dataclass
class IPCError:
    """IPC错误信息"""

    code: str
    message: str
    details: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {"code": self.code, "message": self.message}
        if self.details:
            result["details"] = self.details
        return result

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "IPCError":
        return cls(
            code=data["code"], message=data["message"], details=data.get("details")
        )


@dataclass
class IPCMetadata:
    """IPC元数据"""

    timestamp: str
    request_id: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {"timestamp": self.timestamp}
        if self.request_id:
            result["request_id"] = self.request_id
        return result

    @classmethod
    def create(cls, request_id: Optional[str] = None) -> "IPCMetadata":
        return cls(
            timestamp=datetime.now().isoformat(),
            request_id=request_id or str(uuid.uuid4()),
        )


@dataclass
class IPCResponse:
    """
    纯IPC响应格式 - 完全独立于HTTP概念

    标准格式:
    {
        "success": boolean,
        "data": object | null,
        "error": {
            "code": string,
            "message": string,
            "details": object | null
        } | null,
        "metadata": {
            "timestamp": string,
            "request_id": string | null
        }
    }
    """

    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[IPCError] = None
    metadata: Optional[IPCMetadata] = None

    def __post_init__(self):
        """确保元数据存在"""
        if self.metadata is None:
            self.metadata = IPCMetadata.create()

    @classmethod
    def success_response(
        cls, data: Optional[Dict[str, Any]] = None, request_id: Optional[str] = None
    ) -> "IPCResponse":
        """创建成功响应"""
        return cls(
            success=True, data=data or {}, metadata=IPCMetadata.create(request_id)
        )

    @classmethod
    def error_response(
        cls,
        error_code: Union[str, IPCErrorCode],
        message: str,
        details: Optional[Dict[str, Any]] = None,
        request_id: Optional[str] = None,
    ) -> "IPCResponse":
        """创建错误响应"""
        error_code_str = (
            error_code.value if isinstance(error_code, IPCErrorCode) else error_code
        )

        return cls(
            success=False,
            error=IPCError(code=error_code_str, message=message, details=details),
            metadata=IPCMetadata.create(request_id),
        )

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        result = {
            "success": self.success,
            "data": self.data,
            "metadata": self.metadata.to_dict() if self.metadata else {},
        }

        if self.error:
            result["error"] = self.error.to_dict()
        else:
            result["error"] = None

        return result

    def to_json(self) -> str:
        """转换为JSON字符串"""
        return json.dumps(self.to_dict())

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "IPCResponse":
        """从字典创建响应"""
        error = None
        if data.get("error"):
            error = IPCError.from_dict(data["error"])

        metadata = None
        if data.get("metadata"):
            metadata = IPCMetadata(
                timestamp=data["metadata"]["timestamp"],
                request_id=data["metadata"].get("request_id"),
            )

        return cls(
            success=data["success"],
            data=data.get("data"),
            error=error,
            metadata=metadata,
        )

    @classmethod
    def from_json(cls, json_str: str) -> "IPCResponse":
        """从JSON字符串创建响应"""
        return cls.from_dict(json.loads(json_str))


@dataclass
class IPCRequest:
    """IPC请求对象"""

    method: str
    path: str
    data: Optional[Dict[str, Any]] = None
    query_params: Optional[Dict[str, Any]] = None
    path_params: Optional[Dict[str, str]] = None
    request_id: Optional[str] = None

    def __post_init__(self):
        """确保request_id和headers存在"""
        if self.request_id is None:
            self.request_id = str(uuid.uuid4())
        # 初始化headers属性
        if not hasattr(self, "_headers"):
            self._headers = {}

    @property
    def headers(self) -> Dict[str, str]:
        """获取请求头部"""
        return getattr(self, "_headers", {})

    def get_query(self, name: str, default: Any = None) -> Any:
        """获取查询参数"""
        if not self.query_params:
            return default
        return self.query_params.get(name, default)

    def get_header(self, name: str, default: Any = None) -> Any:
        """获取请求头部信息"""
        # 头部信息由IPC服务器设置在_headers属性中
        headers = getattr(self, "_headers", {})
        return headers.get(name, default)

    def to_dict(self) -> Dict[str, Any]:
        """转换为字典"""
        return {
            "method": self.method,
            "path": self.path,
            "data": self.data,
            "query_params": self.query_params,
            "path_params": self.path_params,
            "request_id": self.request_id,
        }

    def to_json(self) -> str:
        """转换为JSON字符串"""
        return json.dumps(self.to_dict())

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "IPCRequest":
        """从字典创建请求"""
        return cls(
            method=data["method"],
            path=data["path"],
            data=data.get("data"),
            query_params=data.get("query_params"),
            path_params=data.get("path_params"),
            request_id=data.get("request_id"),
        )

    @classmethod
    def from_json(cls, json_str: str) -> "IPCRequest":
        """从JSON字符串创建请求"""
        return cls.from_dict(json.loads(json_str))


# 常用的响应工厂函数


def success_response(
    data: Optional[Dict[str, Any]] = None, request_id: Optional[str] = None
) -> IPCResponse:
    """创建成功响应的便捷函数"""
    return IPCResponse.success_response(data, request_id)


def connection_failed_response(
    message: str, details: Optional[Dict[str, Any]] = None
) -> IPCResponse:
    """连接失败响应"""
    return IPCResponse.error_response(IPCErrorCode.CONNECTION_FAILED, message, details)


def auth_required_response(message: str = "Authentication required") -> IPCResponse:
    """认证要求响应"""
    return IPCResponse.error_response(IPCErrorCode.AUTH_REQUIRED, message)


def invalid_request_response(
    message: str, details: Optional[Dict[str, Any]] = None
) -> IPCResponse:
    """无效请求响应"""
    return IPCResponse.error_response(IPCErrorCode.INVALID_REQUEST, message, details)


def connector_not_found_response(connector_id: str) -> IPCResponse:
    """连接器未找到响应"""
    return IPCResponse.error_response(
        IPCErrorCode.CONNECTOR_NOT_FOUND,
        f"Connector '{connector_id}' not found",
        {"connector_id": connector_id},
    )


def internal_error_response(
    message: str, details: Optional[Dict[str, Any]] = None
) -> IPCResponse:
    """内部错误响应"""
    return IPCResponse.error_response(IPCErrorCode.INTERNAL_ERROR, message, details)


def resource_not_found_response(resource_type: str, resource_id: str) -> IPCResponse:
    """资源未找到响应"""
    return IPCResponse.error_response(
        IPCErrorCode.RESOURCE_NOT_FOUND,
        f"{resource_type} '{resource_id}' not found",
        {"resource_type": resource_type, "resource_id": resource_id},
    )
